File: 2022/07/202207.lua
package.path = "../lua/?.lua;" .. package.path
local ml = require("mikelib")
local split = ml.String.split
local List, Map = ml.List, ml.Map
local unpack = unpack or table.unpack

local function parse_cmd(cmd_str)
    local lines = split(cmd_str, "\n")
    local cmd = split(lines[1], " ")
    if cmd[1] == "ls" then
      local args = lines:sub(2):map(split)
      cmd[2] = args
    end
    return cmd
end

local function parse(input)
  local cmd_strs = split(input, "$ "):sub(2)
  return cmd_strs:map(parse_cmd)
end

local tinput = parse(io.input("tinput.txt"):read("*a"))
local input = parse(io.input("input.txt"):read("*a"))

local orig_tree

-- generate file tree by processing each line in the input
local function generate_tree(input, tree)
  local path = List{tree}
  while #input > 0 do
    local cmd = input:pop()
    if cmd[1] == "cd" then
      local dname = cmd[2]
      if dname == ".." then
        tree = path:pop()
      else
        assert(tree[dname], "Directory does not exist: " .. dname)
        path:append(tree)
        tree = tree[dname]
      end
    elseif cmd[1] == "ls" then
      for _,item in ipairs(cmd[2]) do
        if item[1] == "dir" then
          tree[item[2]] = Map{}
        else
          assert(tonumber(item[1]))
          tree[item[2]] = tonumber(item[1])
        end
      end
    end
  end
end

local function add_sizes(tree)
  local size = 0
  for k,v in pairs(tree) do
    if type(v) == "table" then
      size = size + add_sizes(v)
    else
      size = size + v
    end
  end
  tree.size = size
  return size
end

-- find all directories in the tree that have size at most 100000
local function find_dirs(tree, max_size)
  local dirs = List{}
  for k,v in pairs(tree) do
    if type(v) == "table" then
      if v.size <= max_size then
        dirs:append(List{k, v.size})
      end
      dirs:extend(find_dirs(v, max_size))
    end
  end
  return dirs
end 

local function part1(input)
  local data = input:reversed()
  local tree = Map{["/"]=Map{}}
  generate_tree(data, tree)
  add_sizes(tree)
  local dirs = find_dirs(tree, 100000)
  return dirs:map(function(d) return d[2] end):sum()
end

assert(part1(tinput) == 95437)
print(part1(input))

File: 2022/07/202207.py
def parse_cmd(cstr):
  lines = cstr.splitlines()
  words = lines[0].split()
  if words[0] == 'cd':
    return {'cmd': 'cd', 'args': words[1]}
  elif words[0] == 'ls':
    items = [x.split() for x in lines[1:]]
    return {'cmd': 'ls', 'args': items}

def parse_input(input_str):
  return [parse_cmd(x) for x in input_str.split('$ ')[1:]]

tinput = parse_input(open('tinput.txt', 'r').read())
input = parse_input(open('input.txt', 'r').read())

def find(cmds):
  cmds = list(reversed(cmds))
  nodes = []
  path = []
  while cmds:
    cmd = cmds.pop()
    match cmd:
      case {'cmd': 'cd', 'args': ".."}:
        path.pop()
      case {'cmd': 'cd', 'args': d}:
        path.append(d)
      case {'cmd': 'ls', 'args': items}:
        tsz = 0
        for item in items:
          match item:
            case ['dir', d]:
              pass
            case [sz, nm]:
              tsz += int(sz)
        nodes.append((tuple(path), tsz))
  return nodes

def build_tree(nodes):
  tree = {}
  for path,val in nodes:
    t = tree
    for p in path:
      t.setdefault(p, {})
      t = t[p]
    t['_size'] = t.get('_size', 0) + val
  return tree

def sizes(tree, szs):
  if type(tree) == int:
    return tree
  else:
    sz = sum(sizes(v, szs) for k,v in tree.items())
    szs.append(sz)
    return sz

def part1(cmds):
  nodes = find(cmds)
  tree = build_tree(nodes)
  szs = []
  sizes(tree, szs)
  return sum(s for s in szs if s < 100000)

assert part1(tinput) == 95437

print(part1(input))

def part2(cmds):
  nodes = find(cmds)
  tree = build_tree(nodes)
  szs = []
  sizes(tree, szs)
  szs.sort(reverse=True)
  needed = 30000000 - (70000000 - szs[0])
  big_enough = [x for x in szs if x >= needed]
  return big_enough[-1]

assert part2(tinput) == 24933642
print(part2(input))